`Tutorial groups`: Improve student tutorial group detail page

marlonnienaber opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.


#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
The old tutorial group detail page has room for improvement when it comes to UI/UX.

### Description
<!-- Describe your changes in detail -->
In this PR we refactor the tutorial group detail page that can be accessed through the course overview (not the management page). I built a new version of the page. I fetch the data via a new DTO endpoint (the call of the old page loaded a lot of unnecessary data). The new endpoint (part of TutorialGroupResource )calls a method in TutorialGroupService. While implementing this method some architecture tests started to fail because it is a very large Service and with the new method some threshold regarding Service complexity and size were exceeded. So I looked at the Service and decided one way to improve the complexity a little is to remove the AuthorizationCheckService dependency. I hence moved all authorization related logic into the Resource layer (I personally think that is cleaner anyways).


### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Instructor
- 1 Tutor
- 1 Student
- a course with a tutorial group configuration that enables communication for tutorial groups (see second screenshot if you don't know where to configure this)

1. Create a Tutorial Group with the tutor and register the student
2. Configure it with only the required fields
3. Go to the new tutorial group detail page and verify that the page displays the information you configured correctly:
- next to the title you should find buttons that lead to the tutor chat and to the group chat. Verify that they work
- in the cards some fields should be shown with a "-" since you did not configure all fields for the group.
- all created sessions should be displayed (try both the all sessions and future sessions option for displaying them)
- a disclaimer should be displayed in the attendance card that no data are available 
5. As the instructor edit the group and add the missing fields that you did not configure in the beginning 
6. As the tutor cancel a session, reschedule one and change the location of a third. Also add an attendacne count to some of the sessions.
7. Verify as student that the status of the modified session updates in the new tutorial group detail page. Also verify that the average attendance is now displayed in the attendance card.
8. As the tutor reconfigure the group such that there are no upcoming sessions (change the groups period such that it has already ended)
9. Verify as a student that the next session card in the new tutorial group detail page now displays that there is no upcoming session.


### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->


#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2


### Test Coverage
#### Client
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| course-tutorial-group-detail-container.component.ts | 90.32% | ✅  |
| course-tutorial-group-detail-session-status-chip.component.ts | 100% | ✅ |
| course-tutorial-group-detail-session-status-indicator.component.ts | 100% | ✅ |
| course-tutorial-group-detail.component.ts | 92.53% | ✅  |
| tutorial-groups.service.ts | 87.5% | ✅  |

#### Server
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| TutorialGroupResource.java | 93% | ✅  |
| TutorialGroupService.java | 93% | ✅  |


### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
<img width="1512" height="963" alt="image" src="https://github.com/user-attachments/assets/1e521936-c267-41a8-b1b0-ff72cbff1482" />
<img width="1512" height="982" alt="image" src="https://github.com/user-attachments/assets/71bec174-7d8e-4172-9075-6311f4918a2e" />




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Detailed Tutorial Group view: tutor card, general info, attendance chart, next-session with per-field change indicators, session list with filters, status chips/indicator, loading state, and tutor-chat/group-channel links; client and API support with timezone-aware details and chat creation/navigation.

- **Improvements**
  - Weekday translations unified under global.*, expanded tutorial-group i18n.
  - Profile picture color fallback improved.
  - Removed utilization text from tutorial-group sidebar cards.

- **Tests**
  - Added/updated unit and integration tests for the detail view, timezone handling, and messaging.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tobias-lippert commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/repository/TutorialGroupSessionRepository.java:
```diff
@@ -67,6 +68,21 @@ THEN CONCAT(' - ', CAST(tutorialGroup.campus AS string))
     Set<CalendarEventDTO> getCalendarEventDTOsFromActiveSessionsForTutorialGroupIds(@Param("tutorialGroupIds") Set<Long> tutorialGroupIds);
 
+    @Query("""
+                SELECT new de.tum.cit.aet.artemis.tutorialgroup.util.TutorialGroupDetailSessionData(
```
indentation looks like a bit too much if i'm not mistaken

marlonnienaber replied:
Fixed the indentation.

tobias-lippert commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupSessionResource.java:
```diff
@@ -193,7 +220,13 @@ public ResponseEntity<Void> deleteSession(@PathVariable Long courseId, @PathVari
         checkEntityIdMatchesPathIds(sessionFromDatabase, Optional.of(courseId), Optional.of(tutorialGroupId), Optional.of(sessionId));
-        tutorialGroupService.isAllowedToChangeRegistrationsOfTutorialGroup(sessionFromDatabase.getTutorialGroup(), null);
+
+        User user = userRepository.getUserWithGroupsAndAuthorities();
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        boolean isAdmin = authorizationCheckService.isAdmin(user);
+        boolean isInstructor = authorizationCheckService.isAtLeastInstructorInCourse(course, user);
+        tutorialGroupService.isAllowedToChangeRegistrationsOfTutorialGroup(sessionFromDatabase.getTutorialGroup(), user, isAdmin, isInstructor);
+
```
I think coderabbit has a point here

tobias-lippert commented on src/main/webapp/app/shared/profile-picture/profile-picture.component.ts:
```diff
@@ -13,18 +13,18 @@ import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
     imports: [RouterLink, CommonModule, FontAwesomeModule],
 })
 export class ProfilePictureComponent implements OnInit, OnChanges {
```
Probably some angular dependency update?

tobias-lippert commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail-container/course-tutorial-group-detail-container.component.ts:
```diff
@@ -0,0 +1,70 @@
+        const courseIdParams$ = this.route.parent?.parent?.params;
+        const tutorialGroupIdParams$ = this.route.params;
+        if (courseIdParams$) {
+            this.paramsSubscription = combineLatest([courseIdParams$, tutorialGroupIdParams$])
```
if it is not much effort, could we get rid of the deprecation here?
Other than that I'm not really happy with the code quality in this file but won't comment further on it as it's out of scope for this PR.

tobias-lippert replied:
warning is gone for some reason -> ignore

tobias-lippert commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail-session-status-chip/course-tutorial-group-detail-session-status-chip.component.scss:
```diff
@@ -0,0 +1,23 @@
+    &.status-regular {
+        background-color: var(--green);
+    }
+    &.status-cancelled {
```
I think codacy should be now disabled.

tobias-lippert replied:
@rabeatwork do you know more about it?

marlonnienaber replied:
What is the problem here? I ran stylelint and fixed problems.

tobias-lippert replied:
codacy complained about missing empty line I think. Running stylelint should have fixed that.

tobias-lippert commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail/course-tutorial-group-detail.component.scss:
```diff
@@ -0,0 +1,162 @@
+
+        h1 {
+            font-size: 1.75rem;
+            font-weight: 400;
```
why do we override the h1 style for one specific component. I'm not sure if this is a good idea.

rabeatwork replied:
Yes, please use another h-tag. Also, check that the heading of the tutorial group matches the heading size of an exercise or lecture, for example.

tobias-lippert commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail/course-tutorial-group-detail.component.scss:
```diff
@@ -0,0 +1,162 @@
+
+            h3 {
+                font-size: 1.3rem;
+                margin: 0;
```
same here.
shouldn't you use different h tags then?

tobias-lippert commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail/course-tutorial-group-detail.component.ts:
```diff
@@ -1,70 +1,259 @@
+    tutorChatLink = computed(() => this.computeTutorChatLink(this.course().id, this.tutorialGroup().tutorChatId));
+    groupChannelLink = computed(() => this.computeGroupChannelLink(this.course().id, this.tutorialGroup().groupChannelId));
+
+    readonly faFlag = faFlag;
+    readonly faUsers = faUsers;
+    readonly faTag = faTag;
+    readonly faCircleExclamation = faCircleExclamation;
+    readonly faCalendar = faCalendar;
+    readonly faClock = faClock;
+    readonly faMapPin = faMapPin;
+    readonly faBuildingColumns = faBuildingColumns;
+    readonly faQuestion = faQuestion;
+    readonly faCheck = faCheck;
```
```suggestion
    protected readonly faFlag = faFlag;
    protected readonly faUsers = faUsers;
    protected readonly faTag = faTag;
    protected readonly faCircleExclamation = faCircleExclamation;
    protected readonly faCalendar = faCalendar;
    protected readonly faClock = faClock;
    protected readonly faMapPin = faMapPin;
    protected readonly faBuildingColumns = faBuildingColumns;
    protected readonly faQuestion = faQuestion;
    protected readonly faCheck = faCheck;
   ```
  and please move them to the top, so they can easily be skipped when reviewing/reading the file

tobias-lippert commented on src/main/webapp/app/tutorialgroup/shared/service/tutorial-groups.service.ts:
```diff
@@ -44,6 +45,40 @@ export class TutorialGroupsService {
+                        response.language,
+                        response.isOnline,
+                        (response.sessions ?? []).map(
+                            (session: any) =>
```
more specific type than any?

marlonnienaber replied:
I introduced a RawTutorialGroupDetailGroupDTO and used it instead of any.

tobias-lippert commented on src/test/java/de/tum/cit/aet/artemis/tutorialgroup/AbstractTutorialGroupIntegrationTest.java:
```diff
@@ -304,6 +311,43 @@ ZonedDateTime getDateTimeInBerlinTimeZone(LocalDate date, int hour) {
+        ZonedDateTime firstSessionEnd = ZonedDateTime.of(LocalDate.parse(schedule.getValidFromInclusive()), LocalTime.parse(schedule.getEndTime()),
+                ZoneId.of(course.getTimeZone()));
+
+        // create cancelled session
```
I think the variable name is enough here ;-)

tobias-lippert commented on src/test/javascript/spec/helpers/stubs/tutorialgroup/course-tutorial-group-detail-stub.component.ts:
```diff
@@ -0,0 +1,13 @@
+@Component({
+    selector: 'jhi-course-tutorial-group-detail',
+    template: '',
+    standalone: true,
```
standalone is the default from angular 19 onwards

tobias-lippert reviewed (changes requested):
looks quite good overall.

Left a few inline comments

SamuelRoettgermann reviewed (changes requested):
I don't really like that the save button on the `Edit global Tutorial Groups Configuration` (also a bit inconsistent with the casing) says "Edit". I'd much rather see it being called "Submit" or "Save".

> in the cards some fields should be shown with a "-" since you did not configure all fields for the group.

This does not work for the `Capacity`, `Schedule` and `Meeting Place` fields.

The default time period is always `Saturday, Jun 7, 2025 - Sunday, Jun 8, 2025`. Why not the current time?

> a disclaimer should be displayed in the attendance card that no data are available

This is only visible in the Student View, but maybe that's intended.

> As a tutor edit the group and add the missing fields that you did not configure in the beginning

I was not given the opportunity to edit anything of the tutorial as the tutor.

When I edit a canceled schedule, it automatically gets re-activated (adding a number of attendees does not). Is that intentional/desired behaviour?

From Step 7:
> Also verify that the average attendance is now displayed in the attendance card.

This also didn't work. It only showed me the average attendance in the `Utilization` tab of the tutor's overview, but not in the student view.

> Verify as a student that the next session card in the new tutorial group detail page now displays that there is no upcoming session.

It says "No data available", not "No upcoming session", but I guess that's good enough

Another thing: There is no indication in the student's view whether a schedule is canceled, or at least I couldn't spot it (Can you tell me which of these 4 sessions got cancelled?:) 
<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/ff2c40bb-78f7-4935-8945-8372b3a7a526" />

Answer: It's the 16.06 (second from top) one.


Another thing: I have 2 broken tutorial groups. Namely these are: 
- https://artemis-test4.artemis.cit.tum.de/courses/55/tutorial-groups/8
- https://artemis-test4.artemis.cit.tum.de/courses/55/tutorial-groups/9

where it shows them in the overview, but on clicking them tells me they don't exist. It should be noted that these tutorial groups were created before this PR.

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/repository/TutorialGroupRepository.java:
```diff
@@ -141,6 +142,29 @@ public interface TutorialGroupRepository extends ArtemisJpaRepository<TutorialGr
             """)
     Optional<TutorialGroup> findByIdWithSessions(@Param("tutorialGroupId") long tutorialGroupId);
 
+    @Query("""
+            SELECT new de.tum.cit.aet.artemis.tutorialgroup.util.TutorialGroupDetailGroupData(
+                tutorialGroup.course.id,
+                tutorialGroup.id,
+                tutorialGroup.title,
+                tutorialGroup.language,
+                tutorialGroup.isOnline,
+                tutorialGroup.capacity,
+                tutorialGroup.campus,
+                CONCAT(tutorialGroup.teachingAssistant.firstName, CONCAT(' ', tutorialGroup.teachingAssistant.lastName)),
+                tutorialGroup.teachingAssistant.login,
+                tutorialGroup.teachingAssistant.imageUrl,
+                tutorialGroup.tutorialGroupChannel.id,
+                tutorialGroup.tutorialGroupSchedule.dayOfWeek,
+                tutorialGroup.tutorialGroupSchedule.startTime,
+                tutorialGroup.tutorialGroupSchedule.endTime,
+                tutorialGroup.tutorialGroupSchedule.location
+            )
+            FROM TutorialGroup tutorialGroup
+            WHERE tutorialGroup.id = :tutorialGroupId
+            """)
+    Optional<TutorialGroupDetailGroupData> getTutorialGroupDetailData(@Param("tutorialGroupId") long tutorialGroupId);
```
Each tutorial group is guaranteed to have a teachingAssistant. Left join is therefore only necessary for schedule and channel. Added this.

marlonnienaber commented:
> I don't really like that the save button on the `Edit global Tutorial Groups Configuration` (also a bit inconsistent with the casing) says "Edit". I'd much rather see it being called "Submit" or "Save".
Not implemented by this PR, therefore out of scope.
> > in the cards some fields should be shown with a "-" since you did not configure all fields for the group.
> 
> This does not work for the `Capacity`, `Schedule` and `Meeting Place` fields.
You tested with an existing tutorial group that had capacity already set. Once set it is not possible to remove capacity again. This is a bug in the update logic that is out of scope for this PR. As talked about in person, the rest is intended behavior.
> The default time period is always `Saturday, Jun 7, 2025 - Sunday, Jun 8, 2025`. Why not the current time?
Not implemented by this PR, out of scope.
> > a disclaimer should be displayed in the attendance card that no data are available
> 
> This is only visible in the Student View, but maybe that's intended.
Intended behavior, as this PR only refactors the student view.
> > As a tutor edit the group and add the missing fields that you did not configure in the beginning
> 
> I was not given the opportunity to edit anything of the tutorial as the tutor, aside from the `Sessions` and `Registered Students`.
That was a mistake in the test description. Only instructors are supposed to be able to edit groups.
> When I edit a canceled schedule, it automatically gets re-activated (adding a number of attendees does not). Is that intentional/desired behaviour?
You meant a canceled session. I don't know if it was a design choice, but it is causing no problems. Also not implemented by this PR so out of scope.
> From Step 7:
> 
> > Also verify that the average attendance is now displayed in the attendance card.
> 
> This also didn't work. It only showed me the average attendance in the `Utilization` tab of the tutor's overview, but not in the student view.
The average attendance in the management is a different metric than the attendance displayed in the student view. The Management means the average absolute number of participants, the student view means the average percentage of people that attended with respect to capacity. You did not set capacity for the group but saved the number of attendees for some sessions. Therefore the management can calculate the average attendance but the student view can't because capacity is missing. I will change the naming in the management to avoid confusion and make clear that the two values represnet different metrics.
> > Verify as a student that the next session card in the new tutorial group detail page now displays that there is no upcoming session.
> 
> It says "No data available", not "No upcoming session", but I guess that's good enough
I switched up string keys, thanks for catching that.
> Another thing: There is no indication in the student's view whether a schedule is canceled, or at least I couldn't spot it (Can you tell me which of these 4 sessions got cancelled?:) 
That was indeed a bug. Fixed it.
> Another thing: I have 2 broken tutorial groups. Namely these are:
> 
> * https://artemis-test4.artemis.cit.tum.de/courses/55/tutorial-groups/8
> * https://artemis-test4.artemis.cit.tum.de/courses/55/tutorial-groups/9
The problem was that the groups had no schedule set and my logic could not handle this. I was not aware that this is possible. I adapted my logic such that it can handle this case. The groups work now.

marlonnienaber commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail/course-tutorial-group-detail.component.ts:
```diff
@@ -1,70 +1,259 @@
+            .map((session) => this.computeSessionDataFrom(session));
+    }
+
+    private computeAverageAttendanceRation(sessions: TutorialGroupDetailSessionDTO[], capacity: number | undefined): number | undefined {
+        if (capacity === undefined) return undefined;
+        const sessionsWithAttendance = sessions.filter((session) => session.attendanceCount);
+        if (sessionsWithAttendance.length === 0) return undefined;
+        const averageAttendance = sessionsWithAttendance.reduce((sum, session) => sum + session.attendanceCount!, 0) / sessionsWithAttendance.length;
+        return averageAttendance / capacity;
+    }
```
makes sense. Fixed this.

marlonnienaber commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail/course-tutorial-group-detail.component.ts:
```diff
@@ -5,66 +1,259 @@
 
-    ngOnDestroy(): void {
-        this.paramsSubscription?.unsubscribe();
+    private computePieChartColor(averageAttendanceRatio: number | undefined): Color {
+        if (!averageAttendanceRatio) {
+            return {
+                group: ScaleType.Ordinal,
+                domain: [GraphColors.LIGHT_GREY],
+            } as Color;
+        } else {
+            let color: string | undefined = undefined;
+            if (averageAttendanceRatio >= 0.9) {
+                color = 'var(--red)';
+            } else if (averageAttendanceRatio >= 0.8) {
+                color = 'var(--orange)';
+            } else if (averageAttendanceRatio >= 0.7) {
+                color = 'var(--yellow)';
+            } else {
+                color = 'var(--green)';
+            }
+            return {
+                group: ScaleType.Ordinal,
+                domain: [color, GraphColors.LIGHT_GREY],
+            } as Color;
+        }
+    }
```
No we want green for low and red for high.

marlonnienaber commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail/course-tutorial-group-detail.component.ts:
```diff
@@ -5,66 +1,259 @@
+            'global.weekdays.sunday',
+        ];
+        return keys[weekDayIndex - 1];
+    }
```
fixed this by using .isoWeekday()

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupResource.java:
```diff
@@ -420,7 +458,10 @@ public ResponseEntity<Void> deregisterStudent(@PathVariable Long courseId, @Path
         log.debug("REST request to deregister {} student from tutorial group : {}", studentLogin, tutorialGroupId);
         var tutorialGroupFromDatabase = this.tutorialGroupRepository.findByIdElseThrow(tutorialGroupId);
         var responsibleUser = userRepository.getUserWithGroupsAndAuthorities();
-        tutorialGroupService.isAllowedToChangeRegistrationsOfTutorialGroup(tutorialGroupFromDatabase, responsibleUser);
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        boolean isAdmin = authorizationCheckService.isAdmin(responsibleUser);
+        boolean isInstructor = authorizationCheckService.isAtLeastInstructorInCourse(course, responsibleUser);
+        tutorialGroupService.isAllowedToChangeRegistrationsOfTutorialGroup(tutorialGroupFromDatabase, responsibleUser, isAdmin, isInstructor);
         checkEntityIdMatchesPathIds(tutorialGroupFromDatabase, Optional.of(courseId), Optional.of(tutorialGroupId));
         User studentToDeregister = userRepository.getUserWithGroupsAndAuthorities(studentLogin);
```
done

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/service/TutorialGroupService.java:
```diff
@@ -677,10 +713,12 @@ public void isAllowedToChangeRegistrationsOfTutorialGroup(@NotNull TutorialGroup
      */
-    public void isAllowedToModifySessionsOfTutorialGroup(@NotNull TutorialGroup tutorialGroup, @Nullable User user) {
+    public void isAllowedToModifySessionsOfTutorialGroup(@NotNull TutorialGroup tutorialGroup, @Nullable User user, boolean isAdmin, boolean isInstructor) {
         // ToDo: Clarify if this is the correct permission check
-        if (!this.isAllowedToSeePrivateTutorialGroupInformation(tutorialGroup, user)) {
+        if (!this.isAllowedToSeePrivateTutorialGroupInformation(tutorialGroup, user, isAdmin, isInstructor)) {
             throw new AccessForbiddenException("The user is not allowed to modify the sessions of tutorial group: " + tutorialGroup.getId());
         }
     }
```
done

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/communication/repository/conversation/OneToOneChatRepository.java:
```diff
@@ -75,4 +75,21 @@ AND EXISTS (
 
     Integer countByCreatorIdAndCourseId(Long creatorId, Long courseId);
+
+    @Query("""
```
true.

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/dto/TutorialGroupDetailSessionDTO.java:
```diff
@@ -0,0 +1,49 @@
+     * @param courseTimeZone    the course time zone
+     * @return a DTO with session details and flags for deviations
+     */
+    public static TutorialGroupDetailSessionDTO from(TutorialGroupDetailSessionData data, int scheduleDayOfWeek, LocalTime scheduleStart, LocalTime scheduleEnd,
```
Because when the tutorial group feature was implemented, the design decision was made to save the start and end date of schedules as ISO 8601 formatted strings (HH:mm:ss). These have to be interpreted in the context of the course's timezone. I agree, it is a horrible design decision and I don't know why it was merged like this, but we are stuck with it for now.

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/repository/TutorialGroupSessionRepository.java:
```diff
@@ -67,6 +68,21 @@ THEN CONCAT(' - ', CAST(tutorialGroup.campus AS string))
+                FROM TutorialGroupSession session
+                    JOIN session.tutorialGroup tutorialGroup
+                WHERE tutorialGroup.id = :tutorialGroupId
+                ORDER BY session.start
```
done

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/repository/TutorialGroupRepository.java:
```diff
@@ -141,6 +142,29 @@ public interface TutorialGroupRepository extends ArtemisJpaRepository<TutorialGr
     Optional<TutorialGroup> findByIdWithSessions(@Param("tutorialGroupId") long tutorialGroupId);
 
+    @Query("""
+            SELECT new de.tum.cit.aet.artemis.tutorialgroup.util.TutorialGroupDetailGroupData(
```
Added DTO suffix. See comment above.

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/service/TutorialGroupService.java:
```diff
@@ -571,33 +578,28 @@ private Set<User> tryToFindMatchingUsers(Course course, Set<TutorialGroupRegistr
         tutorialGroups.forEach(tutorialGroup -> this.setTransientPropertiesForUser(user, tutorialGroup));
         tutorialGroups.forEach(tutorialGroup -> {
-            if (!this.isAllowedToSeePrivateTutorialGroupInformation(tutorialGroup, user)) {
+            if (!this.isAllowedToSeePrivateTutorialGroupInformation(tutorialGroup, user, isAdmin, isInstructor)) {
```
done

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupResource.java:
```diff
@@ -202,10 +209,39 @@ public ResponseEntity<TutorialGroup> getOneOfCourse(@PathVariable Long courseId,
+        log.debug("REST request to get tutorial group: {} of course: {}", tutorialGroupId, courseId);
+        var course = courseRepository.findByIdElseThrow(courseId);
+        var user = userRepository.getUserWithGroupsAndAuthorities();
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, user);
```
Ah nice, did not know that existed.

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupResource.java:
```diff
@@ -441,7 +480,10 @@ public ResponseEntity<Void> registerStudent(@PathVariable Long courseId, @PathVa
-        tutorialGroupService.isAllowedToChangeRegistrationsOfTutorialGroup(tutorialGroupFromDatabase, responsibleUser);
+        Course course = courseRepository.findByIdElseThrow(courseId);
+        boolean isAdmin = authorizationCheckService.isAdmin(responsibleUser);
+        boolean isInstructor = authorizationCheckService.isAtLeastInstructorInCourse(course, responsibleUser);
```
done

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupSessionResource.java:
```diff
@@ -103,10 +114,15 @@ public TutorialGroupSessionResource(TutorialGroupSessionRepository tutorialGroup
-        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, session.getTutorialGroup().getCourse(), null);
+        Course course = session.getTutorialGroup().getCourse();
+        authorizationCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, user);
+        boolean isAdmin = authorizationCheckService.isAdmin(user);
```
done

marlonnienaber commented on src/main/java/de/tum/cit/aet/artemis/tutorialgroup/web/TutorialGroupSessionResource.java:
```diff
@@ -130,7 +146,12 @@ public ResponseEntity<TutorialGroupSession> update(@PathVariable Long courseId,
-        tutorialGroupService.isAllowedToModifySessionsOfTutorialGroup(sessionToUpdate.getTutorialGroup(), null);
+
+        User user = userRepository.getUserWithGroupsAndAuthorities();
+        Course course = courseRepository.findByIdElseThrow(courseId);
```
done

marlonnienaber commented on src/main/webapp/app/tutorialgroup/overview/course-tutorial-group-detail-container/course-tutorial-group-detail-container.component.html:
```diff
@@ -0,0 +1,7 @@
+<jhi-loading-indicator-container [isLoading]="(isLoading$ | async) ?? false">
```
done

ekayandan reviewed (commented):
Tested on TS3 and TS2. In general components seem to work (CRUD ops for group and session entities). Only thing I could not do was to see the 'Group Channel' button. Probably I need to configure it in order to see but I don't know how to. Maybe you can add to testing steps.

ekayandan reviewed (dismissed):
Approving manual test, changes work as expected.
Thanks for clarifying the scope of the PR via DM.

SamuelRoettgermann reviewed (approved):
It still shows emptiness for the Language field if a user just entered a blank language, but I guess fixing that issue at your part in the code doesn't really make sense, but rather at the point of editing the course and setting the language.

I'm not personally sure about the striking through of the text of canceled sessions, but I guess it does make it way more obvious that this session won't take place, so I'm fine with it.

Not really sure about the checkmark ("Bestätigungs-Pfeil") in the `Next Session` field when it says "No upcoming session". Maybe a `X` would be more appropriate... but that's just a visual thing that's not super important.

Apart from that, lgtm

marlonnienaber commented:
> Re-tested on TS4:
> 
> It still shows emptiness for the Language field if a user just entered a blank language, but I guess fixing that issue at your part in the code doesn't really make sense, but rather at the point of editing the course and setting the language.
> 
> I'm not personally sure about the striking through of the text of canceled sessions, but I guess it does make it way more obvious that this session won't take place, so I'm fine with it.
> 
> Not really sure about the checkmark ("Bestätigungs-Pfeil") in the `Next Session` field when it says "No upcoming session". Maybe a `X` would be more appropriate... but that's just a visual thing that's not super important.
> 
> Apart from that, lgtm

Those were all design decisions. Thx for retesting.

marlonnienaber commented:
@tobias-lippert see slack chat with @rabeatwork. Lets go with the headings as is for now and generally improve heading use in a follow up.

marlonnienaber commented on supporting_scripts/code-coverage/module-coverage-client/check-client-module-coverage.mjs:
```diff
@@ -49,7 +49,7 @@ const moduleThresholds = {
     },
     core: {
         statements: 89.80,
-        branches:   72.40,
+        branches:   72.30,
         functions:  81.80,
         lines:      89.70,
```
did not add new functionality to the core module.

SamuelRoettgermann reviewed (approved):
Re-tested on TS4. 

Lgtm; I never said it before, but I especially like that the Attendance value can display values above 100% (the Utilization bar in the Manage View can't do that).

Fun fact, the attendance value text can leak into the Attendance circle, but that is such an unrealistic scenario that I think it's fine either way.
<img width="1914" height="1007" alt="image" src="https://github.com/user-attachments/assets/3dee4d46-131a-4f6c-a7f0-7567305697e5" />

tobias-lippert reviewed (approved):
Reapprove code

MoritzSpengler reviewed (approved):
Re-tested on TS4. Detail page looks good. I did not find any problems, thanks for implementing my prior requests.
